<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Credits &mdash; teiphy  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="_static/copybutton.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/clipboard.min.js"></script>
        <script src="_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Command Line Interface Reference" href="cli.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            teiphy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="advanced.html">Advanced Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="example.html">Example</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference.html">API Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command Line Interface Reference</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Credits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#bibliography">Bibliography</a></li>
<li class="toctree-l2"><a class="reference internal" href="#logo">Logo</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">teiphy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Credits</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/credits.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="credits">
<h1>Credits<a class="headerlink" href="#credits" title="Permalink to this heading"></a></h1>
<p><code class="docutils literal notranslate"><span class="pre">teiphy</span></code> was designed by Joey McCollum (Australian Catholic University) and Robert Turnbull (University of Melbourne).
We received additional help from Stephen C. Carlson (Australian Catholic University).</p>
<p>If you use this software, please cite the paper: Joey McCollum and Robert Turnbull, “<code class="docutils literal notranslate"><span class="pre">teiphy</span></code>: A Python Package for Converting TEI XML Collations to NEXUS and Other Formats,” <em>JOSS</em> 7.80 (2022): 4879, DOI: 10.21105/joss.04879.</p>
<div class="highlight-bibtex notranslate"><div class="highlight"><pre><span></span><span class="nc">@article</span><span class="p">{</span><span class="nl">MT2022</span><span class="p">,</span>
<span class="w">    </span><span class="na">author</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Joey McCollum and Robert Turnbull}</span><span class="p">,</span>
<span class="w">    </span><span class="na">title</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{{teiphy: A Python Package for Converting TEI XML Collations to NEXUS and Other Formats}}</span><span class="p">,</span>
<span class="w">    </span><span class="na">journal</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Journal of Open Source Software}</span><span class="p">,</span>
<span class="w">    </span><span class="na">year</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{2022}</span><span class="p">,</span>
<span class="w">    </span><span class="na">volume</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{7}</span><span class="p">,</span>
<span class="w">    </span><span class="na">number</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{80}</span><span class="p">,</span>
<span class="w">    </span><span class="na">pages</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{4879}</span><span class="p">,</span>
<span class="w">    </span><span class="na">publisher</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{The Open Journal}</span><span class="p">,</span>
<span class="w">    </span><span class="na">doi</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{10.21105/joss.04879}</span><span class="p">,</span>
<span class="w">    </span><span class="na">url</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{https://doi.org/10.21105/joss.04879}</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Further details on the capabilities of <code class="docutils literal notranslate"><span class="pre">teiphy</span></code>, particularly in terms of the text-critically valuable features it can map from TEI XML collations to BEAST 2 inputs, are discussed in Joey McCollum and Robert Turnbull, “Using Bayesian Phylogenetics to Infer Manuscript Transmission History,” <em>DSH</em> TBD (2024), DOI: 10.1093/llc/fqad089.</p>
<div class="highlight-bibtex notranslate"><div class="highlight"><pre><span></span><span class="nc">@article</span><span class="p">{</span><span class="nl">MT2024</span><span class="p">,</span>
<span class="w">    </span><span class="na">author</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Joey McCollum and Robert Turnbull}</span><span class="p">,</span>
<span class="w">    </span><span class="na">title</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{{Using Bayesian Phylogenetics to Infer Manuscript Transmission History}}</span><span class="p">,</span>
<span class="w">    </span><span class="na">journal</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{Digital Scholarship in the Humanities}</span><span class="p">,</span>
<span class="w">    </span><span class="na">year</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{2024}</span><span class="p">,</span>
<span class="w">    </span><span class="na">volume</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{TBD}</span><span class="p">,</span>
<span class="w">    </span><span class="na">number</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{TBD}</span><span class="p">,</span>
<span class="w">    </span><span class="na">pages</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{TBD}</span><span class="p">,</span>
<span class="w">    </span><span class="na">doi</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{10.1093/llc/fqad089}</span><span class="p">,</span>
<span class="w">    </span><span class="na">url</span><span class="w"> </span><span class="p">=</span><span class="w"> </span><span class="s">{https://doi.org/10.1093/llc/fqad089}</span>
<span class="p">}</span>
</pre></div>
</div>
<section id="bibliography">
<h2>Bibliography<a class="headerlink" href="#bibliography" title="Permalink to this heading"></a></h2>
<p>teiphy relies on the following research. Please cite as appropriate:</p>
<div class="docutils container" id="id1">
<dl class="citation">
<dt class="label" id="id11"><span class="brackets">AAB+15</span></dt>
<dd><p>Martín Abadi, Ashish Agarwal, Paul Barham, Eugene Brevdo, Zhifeng Chen, Craig Citro, Greg S. Corrado, Andy Davis, Jeffrey Dean, Matthieu Devin, Sanjay Ghemawat, Ian Goodfellow, Andrew Harp, Geoffrey Irving, Michael Isard, Yangqing Jia, Rafal Jozefowicz, Lukasz Kaiser, Manjunath Kudlur, Josh Levenberg, Dandelion Mané, Rajat Monga, Sherry Moore, Derek Murray, Chris Olah, Mike Schuster, Jonathon Shlens, Benoit Steiner, Ilya Sutskever, Kunal Talwar, Paul Tucker, Vincent Vanhoucke, Vijay Vasudevan, Fernanda Viégas, Oriol Vinyals, Pete Warden, Martin Wattenberg, Martin Wicke, Yuan Yu, and Xiaoqiang Zheng. TensorFlow: large-scale machine learning on heterogeneous systems. 2015. URL: <a class="reference external" href="https://www.tensorflow.org/">https://www.tensorflow.org/</a>.</p>
</dd>
<dt class="label" id="id34"><span class="brackets">AAK+14</span></dt>
<dd><p>Barbara Aland, Kurt Aland, Johannes Karavidopoulos, Carlo M. Martini, and Bruce M. Metzger, editors. <em>The Greek New Testament</em>. Deutsche Bibelgesellschaft, Stuttgart, 5 edition, 2014.</p>
</dd>
<dt class="label" id="id15"><span class="brackets">Bal10</span></dt>
<dd><p>Clinton S. Baldwin. Factor analysis: a new method for classifying New Testament Greek manuscripts. <em>Andrews University Seminary Studies</em>, 48(1):29–53, 2010.</p>
</dd>
<dt class="label" id="id23"><span class="brackets">BHBR98</span></dt>
<dd><p>Adrian C. Barbrook, Christopher J. Howe, Norman Blake, and Peter Robinson. The phylogeny of \emph The Canterbury Tales. <em>Nature</em>, 394:839, 1998. <a class="reference external" href="https://doi.org/10.1038/29667">doi:10.1038/29667</a>.</p>
</dd>
<dt class="label" id="id9"><span class="brackets">BVBS+19</span></dt>
<dd><p>Remco Bouckaert, Timothy G. Vaughan, Joëlle Barido-Sottani, Sebastián Duchêne, Mathieu Fourment, Alexandra Gavryushkina, Joseph Heled, Graham Jones, Denise Kühnert, Nicola De Maio, Michael Matschiner, Fábio K. Mendes, Nicola F. Müller, Huw A. Ogilvie, Louis du Plessis, Alex Popinga, Andrew Rambaut, David Rasmussen, Igor Siveroni, Marc A. Suchard, Chieh-Hsi Wu, Dong Xie, Chi Zhang, Tanja Stadler, and Alexei J. Drummond. BEAST 2.5: an advanced software platform for Bayesian evolutionary analysis. <em>PLOS Computational Biology</em>, 15(4):1–28, 2019. <a class="reference external" href="https://doi.org/10.1371/journal.pcbi.1006650">doi:10.1371/journal.pcbi.1006650</a>.</p>
</dd>
<dt class="label" id="id27"><span class="brackets">Car15</span></dt>
<dd><p>Stephen C. Carlson. <em>The Text of Galatians and Its History</em>. Number 385 in Wissenschaftliche Untersuchungen zum Neuen Testament. Mohr Siebeck, Tübingen, 2015. ISBN 978-3-16-153323-5. <a class="reference external" href="https://doi.org/10.1628/978-3-16-153324-2">doi:10.1628/978-3-16-153324-2</a>.</p>
</dd>
<dt class="label" id="id28"><span class="brackets">Edm19</span></dt>
<dd><p>Andrew Charles Edmondson. <em>An Analysis of the Coherence-Based Genealogical Method Using Phylogenetics</em>. University of Birmingham, PhD diss., 2019. URL: <a class="reference external" href="http://etheses.bham.ac.uk/id/eprint/9150">http://etheses.bham.ac.uk/id/eprint/9150</a>.</p>
</dd>
<dt class="label" id="id4"><span class="brackets">Far88</span></dt>
<dd><p>James S. Farris. <em>Hennig86, ver. 1.5. Program and Documentation</em>. James S. Farris, Port Jefferson Station, NY, 1988.</p>
</dd>
<dt class="label" id="id22"><span class="brackets">Fel04</span></dt>
<dd><p>Joseph Felsenstein. <em>Inferring Phylogenies</em>. Sinauer Associates, Sunderland, MA, 2004.</p>
</dd>
<dt class="label" id="id16"><span class="brackets">Fin18</span></dt>
<dd><p>Timothy J. Finney. How to discover textual groups. <em>Digital Studies/Le champ numérique</em>, 2018. <a class="reference external" href="https://doi.org/10.16995/dscn.291">doi:10.16995/dscn.291</a>.</p>
</dd>
<dt class="label" id="id21"><span class="brackets">Fis20</span></dt>
<dd><p>Franz Fischer. Representing the critical text. In Philipp Roelli, editor, <em>Handbook of Stemmatology: History, Methodology, Digital Approaches</em>, De Gruyter Reference, pages 405–427. De Gruyter, Berlin, 2020.</p>
</dd>
<dt class="label" id="id5"><span class="brackets">GC16</span></dt>
<dd><p>Pablo A. Goloboff and Santiago A. Catalano. TNT, version 1.5, including a full implementation of phylogenetic morphometrics. <em>Cladistics</em>, 32(3):221–238, 2016. <a class="reference external" href="https://doi.org/10.1111/cla.12160">doi:10.1111/cla.12160</a>.</p>
</dd>
<dt class="label" id="id31"><span class="brackets">HMvandWalt+20</span></dt>
<dd><p>Charles R. Harris, K. Jarrod Millman, Stéfan J. van der Walt, Ralf Gommers, Pauli Virtanen, David Cournapeau, Eric Wieser, Julian Taylor, Sebastian Berg, Nathaniel J. Smith, Robert Kern, Matti Picus, Stephan Hoyer, Marten H. van Kerkwijk, Matthew Brett, Allan Haldane, Jaime Fernández del Río, Mark Wiebe, Pearu Peterson, Pierre Gérard-Marchant, Kevin Sheppard, Tyler Reddy, Warren Weckesser, Hameer Abbasi, Christoph Gohlke, and Travis E. Oliphant. Array programming with NumPy. <em>Nature</em>, 585:357–362, 2020. <a class="reference external" href="https://doi.org/10.1038/s41586-020-2649-2">doi:10.1038/s41586-020-2649-2</a>.</p>
</dd>
<dt class="label" id="id30"><span class="brackets">Hyytiainen21</span></dt>
<dd><p>Pasi Hyytiäinen. The changing text of Acts: a phylogenetic approach. <em>TC: A Journal of Biblical Textual Criticism</em>, 26:1–28, 2021.</p>
</dd>
<dt class="label" id="id18"><span class="brackets">ISM95</span></dt>
<dd><p>Nancy M. Ide and C.M. Sperberg-McQueen. The TEI: history, goals and future. <em>Computers and the Humanities</em>, 29(1):5–15, 1995. <a class="reference external" href="https://doi.org/10.1007/978-94-011-0325-1_2">doi:10.1007/978-94-011-0325-1_2</a>.</p>
</dd>
<dt class="label" id="id35"><span class="brackets">Lew01</span></dt>
<dd><p>Paul O. Lewis. A likelihood approach to estimating phylogeny from discrete morphological character data. <em>Systematic Biology</em>, 50(6):913–925, November 2001. URL: <a class="reference external" href="https://doi.org/10.1080/106351501753462876">https://doi.org/10.1080/106351501753462876</a> (visited on 2020-11-08), <a class="reference external" href="https://doi.org/10.1080/106351501753462876">doi:10.1080/106351501753462876</a>.</p>
</dd>
<dt class="label" id="id2"><span class="brackets">MSM97</span></dt>
<dd><p>David R. Maddison, David L. Swofford, and Wayne P. Maddison. NEXUS: an extensible file format for systematic information. <em>Systematic Biology</em>, 46(4):590–621, 12 1997. URL: <a class="reference external" href="https://doi.org/10.1093/sysbio/46.4.590">https://doi.org/10.1093/sysbio/46.4.590</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/sysbio/article-pdf/46/4/590/19502018/46-4-590.pdf">arXiv:https://academic.oup.com/sysbio/article-pdf/46/4/590/19502018/46-4-590.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/sysbio/46.4.590">doi:10.1093/sysbio/46.4.590</a>.</p>
</dd>
<dt class="label" id="id6"><span class="brackets">MSC+20</span></dt>
<dd><p>Bui Quang Minh, Heiko A. Schmidt, Olga Chernomor, Dominik Schrempf, Michael D. Woodhams, Arndt von Haeseler, and Robert Lanfear. IQ-TREE 2: new models and efficient methods for phylogenetic inference in the genomic era. <em>Molecular Biology and Evolution</em>, 37(5):1530–1534, 2020. URL: <a class="reference external" href="https://doi.org/10.1093/molbev/msaa015">https://doi.org/10.1093/molbev/msaa015</a>, <a class="reference external" href="https://doi.org/10.1093/molbev/msaa015">doi:10.1093/molbev/msaa015</a>.</p>
</dd>
<dt class="label" id="id10"><span class="brackets">PVG+11</span></dt>
<dd><p>Fabian Pedregosa, Gaël Varoquaux, Alexandre Gramfort, Vincent Michel, Bertrand Thirion, Olivier Grisel, Mathieu Blondel, Peter Prettenhofer, Ron Weiss, Vincent Dubourg, Jake Vanderplas, Alexandre Passos, David Cournapeau, Matthieu Brucher, Matthieu Perrot, and Édouard Duchesnay. Scikit-learn: machine learning in Python. <em>Journal of Machine Learning Research</em>, 12(85):2825–2830, 2011.</p>
</dd>
<dt class="label" id="id20"><span class="brackets">ROHara92</span></dt>
<dd><p>Peter Robinson and Robert J. O'Hara. Report on the Textual Criticism Challenge 1991. <em>Bryn Mawr Classical Review</em>, 3(4):331–337, 1992.</p>
</dd>
<dt class="label" id="id8"><span class="brackets">RTvandMark+12</span></dt>
<dd><p>Fredrik Ronquist, Maxim Teslenko, Paul van der Mark, Daniel L. Ayres, Aaron Darling, Sebastian Höhna, Bret Larget, Liang Liu, Marc A. Suchard, and John P. Huelsenbeck. MRBAYES 3.2: efficient Bayesian phylogenetic inference and model selection across a large model space. <em>Systematic Biology</em>, 61(3):539–542, 2012. <a class="reference external" href="https://doi.org/10.1093/sysbio/sys029">doi:10.1093/sysbio/sys029</a>.</p>
</dd>
<dt class="label" id="id24"><span class="brackets">Sal00</span></dt>
<dd><p>B. J. P. Salemans. <em>Building Stemmas with the Computer in a Cladistic, Neo-Lachmannian, Way: The Case of Fourteen Text Versions of Lanseloet van Denemerken</em>. Katholieke Universiteit Nijmegen, PhD diss., 2000. URL: <a class="reference external" href="https://hdl.handle.net/2066/147058">https://hdl.handle.net/2066/147058</a>.</p>
</dd>
<dt class="label" id="id25"><span class="brackets">SWH02</span></dt>
<dd><p>Matthew Spencer, Klaus Wachtel, and Christopher J. Howe. The Greek Vorlage of the Syra Harclensis: a comparative study on method in exploring textual genealogy. <em>TC: A Journal of Biblical Textual Criticism</em>, 2002. URL: <a class="reference external" href="http://jbtc.org/v07/SWH2002/index.html">http://jbtc.org/v07/SWH2002/index.html</a>.</p>
</dd>
<dt class="label" id="id26"><span class="brackets">SWH04</span></dt>
<dd><p>Matthew Spencer, Klaus Wachtel, and Christopher J. Howe. Representing multiple pathways of textual flow in the Greek manuscripts of the Letter of James using reduced median networks. <em>Computers and the Humanities</em>, 38:1–14, 2004. <a class="reference external" href="https://doi.org/10.1023/B:CHUM.0000009290.14571.59">doi:10.1023/B:CHUM.0000009290.14571.59</a>.</p>
</dd>
<dt class="label" id="id7"><span class="brackets">Sta14</span></dt>
<dd><p>Alexandros Stamatakis. RAxML version 8: a tool for phylogenetic analysis and post-analysis of large phylogenies. <em>Bioinformatics</em>, 30(9):1312–1313, 2014. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btu033">doi:10.1093/bioinformatics/btu033</a>.</p>
</dd>
<dt class="label" id="id3"><span class="brackets">Swo03</span></dt>
<dd><p>D. L. Swofford. Paup*: phylogenetic analysis using parsimony (*and other methods). version 4. 2003.</p>
</dd>
<dt class="label" id="id13"><span class="brackets">Tho02</span></dt>
<dd><p>J. C. Thorpe. Multivariate statistical analysis for manuscript classification. <em>TC: A Journal of Biblical Textual Criticism</em>, 2002. URL: <a class="reference external" href="http://jbtc.org/v07/Thorpe2002.html">http://jbtc.org/v07/Thorpe2002.html</a>.</p>
</dd>
<dt class="label" id="id29"><span class="brackets">Tur20</span></dt>
<dd><p>Robert Turnbull. <em>The Textual History of Codex Sinaiticus Arabicus and Its Family</em>. Ridley College, PhD diss., 2020.</p>
</dd>
<dt class="label" id="id14"><span class="brackets">Wil08</span></dt>
<dd><p>Wieland Willker. Principal component analysis of manuscripts of the Gospel of John. 2008. URL: <a class="reference external" href="http://www.willker.de/wie/TCG/PCA/index.html">http://www.willker.de/wie/TCG/PCA/index.html</a>.</p>
</dd>
<dt class="label" id="id12"><span class="brackets">ZZ12</span></dt>
<dd><p>Marinka Zitnik and Blaz Zupan. NIMFA: a Python library for nonnegative matrix factorization. <em>Journal of Machine Learning Research</em>, 13:849–853, 2012.</p>
</dd>
<dt class="label" id="id17"><span class="brackets">McCollum19</span></dt>
<dd><p>Joey McCollum. Biclustering readings and manuscripts via non-negative matrix factorization, with application to the text of Jude. <em>Andrews University Seminary Studies</em>, 57(1):61–89, 2019.</p>
</dd>
<dt class="label" id="id33"><span class="brackets">McKinney10</span></dt>
<dd><p>Wes McKinney. Data structures for statistical computing in Python. In Stéfan van der Walt and Jarrod Millman, editors, <em>Proceedings of the 9th Python in Science Conference</em>, 56–61. 2010. <a class="reference external" href="https://doi.org/10.25080/Majora-92bf1922-00a">doi:10.25080/Majora-92bf1922-00a</a>.</p>
</dd>
<dt class="label" id="id19"><span class="brackets">TEIConsortium22</span></dt>
<dd><p>TEI Consortium. TEI P5: guidelines for electronic text encoding and interchange: critical apparatus [v.4.4.0]. <a class="reference external" href="https://www.tei-c.org/release/doc/tei-p5-doc/en/html/TC.html">https://www.tei-c.org/release/doc/tei-p5-doc/en/html/TC.html</a>, 2022. [Online; accessed 3-September-2022].</p>
</dd>
<dt class="label" id="id32"><span class="brackets">Thepdteam20</span></dt>
<dd><p>The pandas development team. Pandas-dev/pandas: pandas. 2020. <a class="reference external" href="https://doi.org/10.5281/zenodo.3509134">doi:10.5281/zenodo.3509134</a>.</p>
</dd>
</dl>
</div>
</section>
<section id="logo">
<h2>Logo<a class="headerlink" href="#logo" title="Permalink to this heading"></a></h2>
<p>Elements of the logo are derived from the <a class="reference external" href="https://tei-c.org/about/logos/">TEI logo</a>, licensed under the Creative Commons <a class="reference external" href="https://creativecommons.org/licenses/by/3.0/">CC BY 3.0</a> license.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="cli.html" class="btn btn-neutral float-left" title="Command Line Interface Reference" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Joey McCollum and Robert Turnbull.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>